### PR TITLE
EOS-26732: [IAM Policy] Optional arguments Support for list-policies API

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Problem Statement
- Problem statement
[EOS-26732: Optional arguments Support for list-policies API](https://jts.seagate.com/browse/EOS-26732)

# Design
- Document
[Doc](https://seagate-systems.atlassian.net/wiki/spaces/~232536153/pages/824475714/Optional+Parameter+Support+for+list-policies+API)

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

# Testing Screenshots
1. list-policies without params
image

2. list only attached policies
image

3. list only attached and path = / 
image

4. list policies path=/seagate/rohit/
image

### CODE REVIEW

**Review by shalakadharap — APPROVED** [review]
Looks fine. Just Address codacy issues

**Review by bkirunge7 — APPROVED** [review]

**Comment by rohitkolapkar** [comment]
Pre-Merge Jenkins Job - http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/753/
successful. 
@shalakadharap @bkirunge7

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 1]
```diff
@@ -160,8 +162,15 @@ public ServerResponse create() {
      */
     @Override public ServerResponse list() {
       List<Policy> policyList = null;
+      Map<String, Object> parameters = new HashMap<>();
+      if (requestBody.containsKey("OnlyAttached")) {
+        parameters.put("OnlyAttached", requestBody.get("OnlyAttached"));
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:167]: what parameters are these... for what this is being used. Name of variable does not clear it. Are they query parameters.. what is the purpose of it

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 2]
```diff
@@ -160,8 +162,15 @@ public ServerResponse create() {
      */
     @Override public ServerResponse list() {
       List<Policy> policyList = null;
+      Map<String, Object> parameters = new HashMap<>();
+      if (requestBody.containsKey("OnlyAttached")) {
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:166]: move onlyAttached as constant. If they are being used only inside this function then it should be constant inside this function. Else it should be class level constant

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 3]
```diff
@@ -160,8 +162,15 @@ public ServerResponse create() {
...
+        parameters.put("OnlyAttached", requestBody.get("OnlyAttached"));
+      }
+      if (requestBody.containsKey("PathPrefix")) {
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:169]: PathPrefix should be a constant as this is being used more than once

File: auth/server/src/main/java/com/seagates3/dao/PolicyDAO.java [hunk 4]
```diff
@@ -47,7 +49,9 @@ public Policy find(Account account, String name)
     public
      Policy findByArn(String arn, Account account) throws DataAccessException;
     public
-     List<Policy> findAll(Account account) throws DataAccessException;
+     List<Policy> findAll(
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/dao/PolicyDAO.java:52]: we are changin a public API signature. If this was recently written code and we are sure we have changed all the calling instances then its fine.
If this is a old code, then we should not change the signature as others in opensource community may be using it

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java [hunk 5]
```diff
@@ -19,6 +19,7 @@ Object find(String keyToFind, Object key, Object rootKey,
 
  public
   List<Policy> findAll(String keyToFind, Object key,
+                       Map<String, Object> parameters,
                        String prefix) throws DataAccessException;
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:23]: Signature change for an API is risky. Please check above comment and confirm

File: auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java [hunk 6]
```diff
@@ -148,22 +149,48 @@ Policy find(Object policyArnStr,
...
+                       Object parametersObj) throws DataAccessException {
     Account account = (Account)accountObj;
-    String[] attrs = {
-        LDAPUtils.POLICY_ID,                 LDAPUtils.PATH,
-        LDAPUtils.POLICY_CREATE_DATE,        LDAPUtils.POLICY_UPDATE_DATE,
-        LDAPUtils.DEFAULT_VERSION_ID,        LDAPUtils.POLICY_DOC,
-        LDAPUtils.POLICY_NAME,               LDAPUtils.IS_POLICY_ATTACHABLE,
-        LDAPUtils.POLICY_ARN,                LDAPUtils.POLICY_ATTACHMENT_COUNT,
-        LDAPUtils.POLICY_PERMISSION_BOUNDARY};
+    Map<String, Object> parameters = (Map<String, Object>)parametersObj;
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java:155]: parameter is a very generic term. We should avoid it

File: auth/server/src/main/java/com/seagates3/parameter/validator/AbstractParameterValidator.java [hunk 7]
```diff
@@ -82,6 +82,11 @@ public Boolean isValidListParams(Map<String, String> requestBody) {
                     requestBody.get("Marker"));
         }
 
+        if (requestBody.containsKey("OnlyAttached")) {
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/parameter/validator/AbstractParameterValidator.java:85]: do we have a global static constant file. I can see only attached and other keys are used frequently which can be move there

**Review by kupranay — CHANGES_REQUESTED** [review]

File: auth/server/src/main/java/com/seagates3/dao/PolicyDAO.java (hunk 4 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/dao/PolicyDAO.java:52]: @kupranay , This is recently written code. and I have changed all the calling instances.

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java (hunk 5 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:23]: This is recently written code. and I have changed all the calling instances.

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 1 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java]: fixed

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 2 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java]: fixed

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 3 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java]: fixed

File: auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java (hunk 6 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java]: fixed

File: auth/server/src/main/java/com/seagates3/parameter/validator/AbstractParameterValidator.java (hunk 7 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/parameter/validator/AbstractParameterValidator.java:85]: fixed

File: auth/server/src/main/java/com/seagates3/dao/ldap/MotrStore.java [hunk 8]
```diff
@@ -21,7 +21,8 @@
   }
 
   @Override public List<Policy> findAll(
-      String keyToFind, Object obj, String prefix) throws DataAccessException {
+      String keyToFind, Object obj, Map<String, Object> apiParameters,
+      String prefix) throws DataAccessException {
     // TODO Auto-generated method stub
```
Comment (kupranay) [inline auth/server/src/main/java/com/seagates3/dao/ldap/MotrStore.java:26]: Remove TODO Tag

**Review by kupranay — APPROVED** [review]

**Review by shalakadharap — APPROVED** [review]

File: auth/server/src/main/java/com/seagates3/dao/ldap/MotrStore.java (hunk 8 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/dao/ldap/MotrStore.java]: fixed @kupranay . please check

**Comment by rohitkolapkar** [comment]
Pre-merge successful 
http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/764/

**Pull request merged**